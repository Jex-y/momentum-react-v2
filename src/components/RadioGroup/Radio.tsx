import React, { FC, useContext, useRef } from 'react';
import { FocusRing } from '@react-aria/focus';
import { useRadio } from '@react-aria/radio';

import classnames from 'classnames';

import { STYLE, DEFAULTS } from './RadioGroup.constants';
import { RadioProps } from './RadioGroup.types';
import { RadioContext } from './RadioGroup';

import './RadioGroup.style.scss';

const Radio: FC<RadioProps> = (props: RadioProps) => {
  const { className, isDisabled = DEFAULTS.OPTION_DISABLED, id, style, label } = props;
  const state = useContext(RadioContext);
  const ref = useRef(null);
  const { inputProps } = useRadio({ 'aria-label': label, ...props }, state, ref);
  return (
    <label
      data-disabled={isDisabled}
      className={classnames(STYLE.wrapper, className)}
      style={style}
      id={id}
    >
      <FocusRing>
        <input {...inputProps} className={STYLE.button} ref={ref} />
      </FocusRing>
      {label}
    </label>
  );
};

export default Radio;
