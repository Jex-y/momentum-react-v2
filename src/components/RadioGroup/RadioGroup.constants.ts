const CLASS_PREFIX = 'md-radio-group';

const DEFAULTS = {};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
