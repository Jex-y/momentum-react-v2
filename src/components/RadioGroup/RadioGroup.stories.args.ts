import { commonStyles } from '../../storybook/helper.stories.argtypes';

const radioGroupArgTypes = {
  // NOTE: Below is an example. See [Storybook argTypes documentation]{@link https://storybook.js.org/docs/react/api/argtypes}.
  children: {
    description: 'Provides the child nodes for this element.', // NOTE: Description of this prop.
    control: { type: 'text' }, // NOTE: Control type for this prop.
    table: {
      disable: true,
    },
  },
};

export { radioGroupArgTypes };

export default {
  ...commonStyles,
  ...radioGroupArgTypes,
};
