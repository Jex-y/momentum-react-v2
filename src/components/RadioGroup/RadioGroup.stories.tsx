import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';
import React from 'react';

import { Radio, RadioGroup, RadioGroupProps } from './';
import argTypes from './RadioGroup.stories.args';
import Documentation from './RadioGroup.stories.docs.mdx';
import { action } from '@storybook/addon-actions';

export default {
  title: 'Momentum UI/RadioGroup',
  component: RadioGroup,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
  args: {
    onChange: action('onChange'),
  },
};

// NOTE: Primary story. This renders a single component with all external props.
const Example = Template<RadioGroupProps>(RadioGroup).bind({});

Example.argTypes = { ...argTypes };

// TODO: Provide default arguments for this story here. These populate into the argument table for this component.
Example.args = {
  children: (
    <>
      <Radio value="option1">Option 1</Radio>
      <Radio value="option2">Option 2</Radio>
      <Radio value="option3">Option 3</Radio>
    </>
  ),
  label: 'Example',
};

// TODO: Inject additional stories here.

// NOTE: Common variants story. This renders multiple variants of a single component.
const Common = MultiTemplate<RadioGroupProps>(RadioGroup).bind({});

Common.argTypes = { ...argTypes };
delete Common.argTypes.children;

// TODO: Provide default arguments for this story here. These populate into the argument table for this component for all variants.
Common.args = {};

Common.parameters = {
  variants: [
    {
      children: (
        <>
          <Radio value="option1">Selected</Radio>
        </>
      ),
      value: 'option1',
    },
    {
      children: (
        <>
          <Radio value="option1">Unselected</Radio>
        </>
      ),
      value: 'option2',
    },
  ],
};

// NOTE: Export stories here. The first export should be `Example`, and the last export should be `Common`.
export { Example, Common };
