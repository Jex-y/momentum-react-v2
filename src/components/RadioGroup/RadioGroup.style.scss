.md-radio-group {
  > .md-radio-wrapper {
    display: flex;
    align-items: center;
    width: fit-content;
    font-size: 1rem;

    > .md-radio-selected,
    .md-radio-not-selected {
      height: 1rem;
      width: 1rem;
      border-radius: 0.5rem;
      outline: none !important;
      display: flex;
      flex-direction: row;
      justify-content: center;
      align-items: center;
      margin-right: 0.5rem;
      margin-left: 0.5rem;
    }

    &[data-disabled='false'] {
      .md-radio-icon {
        svg {
          fill: var(--radio-checked-icon);
        }
      }

      > .md-radio-label {
        color: var(--radio-unchecked-text);
      }

      > .md-radio-selected {
        background-color: var(--radio-checked-background);
      }

      > .md-radio-not-selected {
        background-color: var(--radio-unchecked-background);
        border-style: solid;
        border-color: var(--radio-unchecked-border-color);
        border-width: 0.0625rem;
      }

      &:hover,
      &.hover {
        color: var(--radio-checked-text-hovered);

        > .md-radio-selected {
          background-color: var(--radio-checked-background-hovered);
        }

        > .md-radio-not-selected {
          background-color: var(--radio-unchecked-background-hovered);
        }
      }

      &:active,
      &.active {
        color: var(--radio-unchecked-text-pressed);

        > .md-radio-selected {
          background-color: var(--radio-checked-background-pressed);
        }

        > .md-radio-not-selected {
          background-color: var(--radio-unchecked-background-pressed);
        }
      }

      .md-radio-focus {
        box-shadow: var(--md-globals-focus-ring-box-shadow-accent);
      }

      &.focus {
        .radio {
          box-shadow: var(--md-globals-focus-ring-box-shadow-accent);
        }
      }
    }

    &[data-disabled='true'],
    &.disable {
      .md-radio-icon {
        svg {
          fill: var(--radio-checked-icon-disabled);
        }
      }

      color: var(--radio-unchecked-text-disabled);

      > .md-radio-selected {
        background-color: var(--radio-checked-background-disabled);
        border-style: none;
      }

      > .md-radio-not-selected {
        background-color: var(--radio-unchecked-background-disabled);
        border-style: none;
      }
    }
  }
}
