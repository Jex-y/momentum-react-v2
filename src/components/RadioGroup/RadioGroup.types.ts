import { CSSProperties, ReactElement, ReactNode } from 'react';
import { RadioAriaProps } from '@react-aria/radio';

export interface GroupProps {
  /**
   * Child components of this RadioGroup.
   */
  children?: ReactNode;

  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * Custom id for overriding this component's CSS.
   */
  id?: string;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;

  label?: string;
}

export interface Props extends RadioAriaProps {
  style?: CSSProperties;
}

export type RadioValue = {
  value: string;
  child: ReactNode;
};
