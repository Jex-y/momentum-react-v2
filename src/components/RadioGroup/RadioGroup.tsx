import React, { FC, ReactElement, useContext, useRef } from 'react';
import classnames from 'classnames';
import { useRadioGroup, useRadio } from '@react-aria/radio';
import { useRadioGroupState } from '@react-stately/radio';

import { STYLE } from './RadioGroup.constants';
import { GroupProps, Props, RadioValue } from './RadioGroup.types';
import './RadioGroup.style.scss';

// TODO: Update JSDOC for this component.
/**
 * The RadioGroup component.
 */

export const RadioContext = React.createContext(null);

const RadioGroup: FC<GroupProps> = (props: GroupProps) => {
  const { className, label, children, id, style } = props;

  const state = useRadioGroupState(props);
  const { radioGroupProps, labelProps } = useRadioGroup(props, state);

  let childElement: ReactElement;
  if (React.isValidElement(children)) {
    childElement = children;
  } else if (Array.isArray(children)) {
    childElement = (
      <>
        {children.map((child: string | RadioValue) => {
          if (typeof child === 'string') {
            return <Radio value={child}>{child}</Radio>;
          } else {
            return <Radio value={child.value}>{child.child}</Radio>;
          }
        })}
      </>
    );
  }

  return (
    <div
      {...radioGroupProps}
      className={classnames(className, STYLE.wrapper)}
      id={id}
      style={style}
    >
      <span {...labelProps}>{label}</span>
      <RadioContext.Provider value={state}>{childElement}</RadioContext.Provider>
    </div>
  );
};

const Radio: FC<Props> = (props: Props) => {
  const { children } = props;
  const state = useContext(RadioContext);
  const ref = useRef(null);
  const { inputProps } = useRadio(props, state, ref);
  return (
    <label style={{ display: 'block' }}>
      <input {...inputProps} ref={ref} />
      {children}
    </label>
  );
};

export default RadioGroup;
export { RadioGroup, Radio };
