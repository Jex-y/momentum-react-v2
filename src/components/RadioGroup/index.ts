import { RadioGroup, Radio } from './RadioGroup';
import * as CONSTANTS from './RadioGroup.constants';
import { Props, GroupProps, RadioValue as _RadioValue } from './RadioGroup.types';

export { CONSTANTS as RADIO_GROUP_CONSTANTS };

export type RadioProps = Props;
export type RadioGroupProps = GroupProps;
export type RadioValue = _RadioValue;
export { RadioGroup, Radio };
export default RadioGroup;
