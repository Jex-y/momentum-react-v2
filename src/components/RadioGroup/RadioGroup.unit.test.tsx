import React from 'react';
import { mount } from 'enzyme';

import RadioGroup, { RADIO_GROUP_CONSTANTS as CONSTANTS } from './';

describe('<RadioGroup />', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<RadioGroup />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(<RadioGroup className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mount(<RadioGroup id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<RadioGroup style={style} />);

      expect(container).toMatchSnapshot();
    });

    /* ...additional snapshot tests... */
  });

  describe('attributes', () => {
    it('should have its wrapper class', () => {
      expect.assertions(1);

      const element = mount(<RadioGroup />)
        .find(RadioGroup)
        .getDOMNode();

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = mount(<RadioGroup className={className} />)
        .find(RadioGroup)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';

      const element = mount(<RadioGroup id={id} />)
        .find(RadioGroup)
        .getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = mount(<RadioGroup style={style} />)
        .find(RadioGroup)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    /* ...additional attribute tests... */
  });

  describe('actions', () => {
    /* ...action tests... */
  });
});
