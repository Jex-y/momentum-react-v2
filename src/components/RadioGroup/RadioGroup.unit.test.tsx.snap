// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> snapshot should match snapshot 1`] = `
<RadioGroup>
  <div
    aria-orientation="vertical"
    className="md-radio-group-wrapper"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <span />
  </div>
</RadioGroup>
`;

exports[`<RadioGroup /> snapshot should match snapshot with className 1`] = `
<RadioGroup
  className="example-class"
>
  <div
    aria-orientation="vertical"
    className="example-class md-radio-group-wrapper"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <span />
  </div>
</RadioGroup>
`;

exports[`<RadioGroup /> snapshot should match snapshot with id 1`] = `
<RadioGroup
  id="example-id"
>
  <div
    aria-orientation="vertical"
    className="md-radio-group-wrapper"
    id="example-id"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="radiogroup"
  >
    <span />
  </div>
</RadioGroup>
`;

exports[`<RadioGroup /> snapshot should match snapshot with style 1`] = `
<RadioGroup
  style={
    Object {
      "color": "pink",
    }
  }
>
  <div
    aria-orientation="vertical"
    className="md-radio-group-wrapper"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="radiogroup"
    style={
      Object {
        "color": "pink",
      }
    }
  >
    <span />
  </div>
</RadioGroup>
`;
